jax/_src/xla_bridge.py

```diff
@@ -519,25 +520,52 @@ def backends() -> dict[str, xla_client.Client]:
           raise RuntimeError(err_msg)
 
     assert _default_backend is not None
-    # We don't warn about falling back to CPU on Mac OS, because we don't
-    # support anything else there at the moment and warning would be pointless.
-    if (py_platform.system() != "Darwin" and
-        _default_backend.platform == "cpu" and
-        _PLATFORM_NAME.value != 'cpu'):
-      logger.warning('No GPU/TPU found, falling back to CPU. '
-                      '(Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)')
+    if not config.jax_platforms:
+      _suggest_missing_backends()
     return _backends
 
 
+def _suggest_missing_backends():
+  if py_platform.system() != "Linux":
+    # If you're not using Linux (or WSL2), we don't have any suggestions at the
+    # moment.
+    return
+
+  default_platform = _default_backend.platform
+  nvidia_gpu_devices = [
+    "/dev/nvidia0",
+    "/dev/dxg",  # WSL2
+  ]
+  google_tpu_devices = [
+    "/dev/accel0",  # TPU v4 or earlier
+    "/dev/vfio/0",  # TPU v5
+  ]
+  if ("cuda" not in _backends and
+      any(os.path.exists(d) for d in nvidia_gpu_devices)):
+    if jax._src.lib.gpu_solver is not None:
+      err = _backend_errors["cuda"]
+      logger.warning(f"CUDA backend failed to initialize: {err} (Set "
+                     "TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)")
+    else:
+      logger.warning("An NVIDIA GPU may be present on this machine, but a "
+                     "CUDA-enabled jaxlib is not installed. Falling back to "
+                     f"{default_platform}.")
+  elif ("tpu" not in _backends and
+        any (os.path.exists(d) for d in google_tpu_devices)):
+    logger.warning("A Google TPU may be present on this machine, but a "
+                    "TPU-enabled jaxlib is not installed. Falling back to "
+                    f"{default_platform}.")
```

Comment:
I'm not sure about the TPU message. I think it's more likely that libtpu isn't installed, or the vfio device is something else.

I wonder if we should always print a message saying which platform is running? Instead of default no message, and trying to message if we think you made a mistake. Log spam is the worst, but jax new users often ask how they know if their code is running on the gpu/tpu (answer: it should say nothing). I also wonder if always printing the platform will be useful with 3p plugins.

I'm not sure about this. We could also just change this message, or not do anything for TPU yet (I think it's rare for people to fall back to CPU on a TPU machine, usually libtpu is installed so it will at least error). Maybe worth polling a broader audience?

Reply:
Note that folks were previously annoyed by the "No GPU" log message on CPU. I think therefore that "always logging" is not great.

How about I just omit the vfio device for now? I agree it'd possibly have false positives, and it's probably better to say nothing in that case.

Reply:
Who was annoyed? My untested hypothesis is that always printing the platform will be less annoying than printing "No GPU/TPU found", because it doesn't look an error or mistake. But still agree logging sucks.

Omitting sounds ok. It's a bit dicey because we're making it even harder to tell that you're falling back to CPU on TPU v5, although not sure anyone notices the logging anyway. Please also omit the part about TPU-enabled jaxlib not being installed, I think libtpu missing is more likely. I would say ", but could not initialize TPU runtime, perhaps due to missing libtpu package."

Reply:
PTAL.

I added reliable detection of TPUs using PCI vendor IDs (cf https://github.com/pytorch/xla/blob/a0a3aba9e5263e5aafd6e7895f36e2703ed3c4cd/torch_xla/_internal/tpu.py#L94-L107).

This code will not produce false positives.

I also reworded the text.

Reply:
See also https://github.com/google/jax/pull/10272 for other discussion